File: server/src/controllers/analysisController.js
import axios from "axios";
import { GoogleGenerativeAI } from "@google/generative-ai";
import { generateFeedbackPrompt } from "../utils/generateFeedbackPrompt.js";

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;

if (!GEMINI_API_KEY) {
  throw new Error("GEMINI_API_KEY is not set in environment variables");
}

const genAI = new GoogleGenerativeAI(GEMINI_API_KEY);

export const transcribeMedia = async (req, res) => {
  try {
    const { mediaUrl, audioData } = req.body;

    let audioBuffer;
    let mimeType = "audio/webm";

    // If audioData is provided directly (from frontend blob), use it (faster)
    if (audioData) {
      audioBuffer = Buffer.from(audioData, "base64");
    } else if (mediaUrl) {
      // Fallback: download from Cloudinary URL
      const response = await axios.get(mediaUrl, {
        responseType: "arraybuffer",
      });
      audioBuffer = Buffer.from(response.data);
      // Detect mime type from URL
      if (mediaUrl.includes(".mp4") || mediaUrl.includes("video")) {
        mimeType = "video/webm";
      }
    } else {
      return res.status(400).json({ message: "mediaUrl or audioData is required" });
    }

    const model = genAI.getGenerativeModel({
      model: "gemini-2.5-flash-lite",
    });

    // Optimized prompt for faster response
    const result = await model.generateContent([
      {
        text: "Transcribe this audio. Return only the transcript text, nothing else.",
      },
      {
        inlineData: {
          mimeType,
          data: audioBuffer.toString("base64"),
        },
      },
    ]);

    const transcript = result.response.text();

    return res.status(200).json({ transcript });
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Transcription error", error);
    
    // Check for rate limit errors
    if (error.status === 429) {
      return res.status(429).json({ 
        message: "Rate limit exceeded. Please wait and try again.", 
        error: String(error) 
      });
    }
    
    return res
      .status(500)
      .json({ 
        message: "Failed to transcribe media", 
        error: String(error),
        details: error.message 
      });
  }
};

export const generateFeedback = async (req, res) => {
  try {
    const {
      transcript,
      durationSeconds,
      topic,
    } = req.body;

    if (transcript === undefined || durationSeconds === undefined || !topic) {
      return res.status(400).json({
        message: "transcript, durationSeconds, and topic are required",
      });
    }

    const prompt = generateFeedbackPrompt(transcript, durationSeconds, topic);

    // Retry logic for API overload
    let result;
    let attempts = 0;
    const maxAttempts = 3;
    
    while (attempts < maxAttempts) {
      try {
        const model = genAI.getGenerativeModel({
          model: "gemini-2.5-flash-lite",
          generationConfig: {
            temperature: 0.3,
            maxOutputTokens: 500,
          },
        });

        result = await model.generateContent(prompt);
        break; // Success, exit loop
      } catch (error) {
        attempts++;
        if (error.status === 503 && attempts < maxAttempts) {
          console.log(`API overloaded, retrying... (${attempts}/${maxAttempts})`);
          await new Promise(resolve => setTimeout(resolve, 1000 * attempts)); // Exponential backoff
        } else {
          throw error; // Re-throw if not 503 or max attempts reached
        }
      }
    }

    let raw = result.response.text() ?? "{}";

    // Some models may wrap JSON in markdown fences like ```json ... ```
    raw = raw.trim();

    // Prefer regex to strip fenced blocks if present
    const fenceMatch = raw.match(/```(?:json)?\s*([\s\S]*?)```/i);
    if (fenceMatch && fenceMatch[1]) {
      raw = fenceMatch[1].trim();
    } else if (raw.startsWith("```")) {
      // Fallback: slice from first { to last }
      const firstBrace = raw.indexOf("{");
      const lastBrace = raw.lastIndexOf("}");
      if (firstBrace !== -1 && lastBrace !== -1) {
        raw = raw.slice(firstBrace, lastBrace + 1);
      }
    }

    let parsed;
    try {
      parsed = JSON.parse(raw);
    } catch (parseError) {
      // eslint-disable-next-line no-console
      console.error("Failed to parse feedback JSON", parseError, raw);
      return res.status(500).json({
        message: "Failed to parse feedback JSON",
      });
    }

    return res.status(200).json(parsed);
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Feedback generation error", error);
    return res.status(500).json({
      message: "Failed to generate feedback",
      error: String(error),
    });
  }
};

File: backend/src/controllers/analysis_controller.py
import asyncio
import base64
import json
import os
import re
from typing import Optional

import google.generativeai as genai
import httpx
from fastapi import HTTPException
from google.api_core.exceptions import ResourceExhausted, ServiceUnavailable
from pydantic import BaseModel

from ..utils.generate_feedback_prompt import generate_feedback_prompt

GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
if not GEMINI_API_KEY:
    raise RuntimeError("GEMINI_API_KEY is not set in environment variables")

genai.configure(api_key=GEMINI_API_KEY)

MODEL_NAME = "gemini-2.5-flash-lite"


class TranscribeBody(BaseModel):
    mediaUrl: Optional[str] = None
    audioData: Optional[str] = None


class FeedbackBody(BaseModel):
    transcript: Optional[str] = None
    durationSeconds: Optional[float] = None
    topic: Optional[str] = None


async def transcribe_media(body: TranscribeBody) -> dict:
    try:
        mime_type = "audio/webm"

        if body.audioData:
            audio_bytes = base64.b64decode(body.audioData)
        elif body.mediaUrl:
            async with httpx.AsyncClient(timeout=60.0) as client:
                resp = await client.get(body.mediaUrl)
                resp.raise_for_status()
                audio_bytes = resp.content
            if ".mp4" in body.mediaUrl or "video" in body.mediaUrl:
                mime_type = "video/webm"
        else:
            raise HTTPException(
                status_code=400, detail={"message": "mediaUrl or audioData is required"}
            )

        model = genai.GenerativeModel(MODEL_NAME)
        result = await model.generate_content_async(
            [
                "Transcribe this audio. Return only the transcript text, nothing else.",
                {"mime_type": mime_type, "data": audio_bytes},
            ]
        )

        return {"transcript": result.text}
    except HTTPException:
        raise
    except ResourceExhausted as error:
        raise HTTPException(
            status_code=429,
            detail={"message": "Rate limit exceeded. Please wait and try again.", "error": str(error)},
        )
    except Exception as error:
        print("Transcription error", error)
        raise HTTPException(
            status_code=500,
            detail={
                "message": "Failed to transcribe media",
                "error": str(error),
                "details": str(error),
            },
        )


def _strip_json_fences(raw: str) -> str:
    raw = raw.strip()
    fence_match = re.search(r"```(?:json)?\s*([\s\S]*?)```", raw, re.IGNORECASE)
    if fence_match:
        return fence_match.group(1).strip()
    if raw.startswith("```"):
        first_brace = raw.find("{")
        last_brace = raw.rfind("}")
        if first_brace != -1 and last_brace != -1:
            return raw[first_brace : last_brace + 1]
    return raw


async def generate_feedback(body: FeedbackBody) -> dict:
    if body.transcript is None or body.durationSeconds is None or not body.topic:
        raise HTTPException(
            status_code=400,
            detail={"message": "transcript, durationSeconds, and topic are required"},
        )

    prompt = generate_feedback_prompt(body.transcript, body.durationSeconds, body.topic)

    try:
        result = None
        max_attempts = 3
        for attempt in range(1, max_attempts + 1):
            try:
                model = genai.GenerativeModel(
                    MODEL_NAME,
                    generation_config=genai.types.GenerationConfig(
                        temperature=0.3, max_output_tokens=500
                    ),
                )
                result = await model.generate_content_async(prompt)
                break
            except ServiceUnavailable:
                if attempt < max_attempts:
                    print(f"API overloaded, retrying... ({attempt}/{max_attempts})")
                    await asyncio.sleep(1 * attempt)
                else:
                    raise

        raw = _strip_json_fences(result.text or "{}")

        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError as parse_error:
            print("Failed to parse feedback JSON", parse_error, raw)
            raise HTTPException(status_code=500, detail={"message": "Failed to parse feedback JSON"})

        return parsed
    except HTTPException:
        raise
    except Exception as error:
        print("Feedback generation error", error)
        raise HTTPException(
            status_code=500, detail={"message": "Failed to generate feedback", "error": str(error)}
        )
